### ISSUE TITLE
Upgrade support for Spring Data Neo4J.

### ISSUE THREAD

**Issue by mangrish** [spring-projects/spring-boot#7391]
**Spring Data Neo4J (SDN)** has had some minor configuration changes made from version 4.1 to 4.2 in order to integrate more tightly with the Spring Data ecosystem.

As of **SDN 4.2.x**, **_Spring Boot will not work with it out of the box._**

An example of how the new changes look can be seen in this example project: 
https://github.com/neo4j-examples/sdn4-university/tree/4.2.x

A blog on how to use SDN 4.2 with Spring Boot 1.4.1 can also be found here: http://graphaware.com/neo4j/2016/09/30/upgrading-to-sdn-42.html

We would love to make sure that when Spring Data Ingalls goes GA, we have the support in Spring Boot to transition users across.

We are able to help you do this work. Please let me know what's the best way to move forward with this.

-Mark

**Comment by philwebb** [comment]
See also #6709

**Comment by wilkinsona** [comment]
And #6976

**Comment by wilkinsona** [comment]
[Here's a commit](https://github.com/wilkinsona/spring-boot/commit/21125d552c94ac47238c8a4fcc1924a61bf6df71) that upgrades to Ingalls snapshots (Spring Data Neo4j 4.2 snapshots) and OGM 2.1 snapshots. It appears to work quite well -- our sample now works both in an IDE and from an executable jar -- but it does raise a few questions:

1. A `Neo4jOperations` bean is no longer available which may break some users when upgrading. Should we auto-configure one even though it's deprecated?
2. Configuring a scope doesn't appear to be supported any more so I've removed the property metadata. Was that the right thing to do?
3. Some of the SDN class names appear to be inconsistent. Some use `Graph` and some use `Neo4j` as the prefix. Are these going to change? My understanding from @olivergierke was that `Neo4j` was to be preferred.

@mangrish Can you help with the above questions please?

**Comment by mangrish** [comment]
1. We are still thinking whether to deprecate `Neo4jOperations`/`Neo4jTemplate`. We feel it may be going the way of `JpaTemplate` and `HibernateTemplate` in that it is superfluous because users can simply autowire a `Session` instead. I will discuss with our team and get back to you on this.

2. Yes. Any reference to the wiring up of a session through the configuration should no longer appear (and correctly is removed in the commit). The sample app and blog post show the correct way to configure SDN in various configurations (e.e. remembering to wire up an interceptor when an MVC app is autodetected for example). I don't see this sort of conditional configuration for that scenario at the moment. Is there anything you need from us to get that working too?

3. Yes, all internally configured beans should have the `Neo4j` prefix. We have left `GraphRepository` as is because we don't want upgrading users to have to change all their repositories. I see there are a few classes still remaining with the old `Graph` prefix. AFAIK they don't affect the boot configuration. If they do let me know which ones otherwise I'll bring it up with the team regarding the other ones.

**Comment by wilkinsona** [comment]
I missed the `OpenSessionInViewInterceptor` in the blog post. Thanks for pointing it out. Spring Boot 2.0 will stop configuring `OpenEntityManagerInViewInterceptor` by default (see #7125). With that in mind, auto-configuration of SDN's `OpenSessionInViewInterceptor` should be opt-in too.

**Comment by wilkinsona** [comment]
[Here's an updated commit](https://github.com/wilkinsona/spring-boot/commit/47587e6aec00da38aa64935e171c7ee0683108b2) that auto-configures `OpenSessionInViewInterceptor` when the user opts in.

I can't see any references to any `Graph…` classes other than `GraphRepository` in our auto-configuration. Our sample uses `GraphId` but I imagine its name isn't changing for the same reason as `GraphRepository`.

In summary, I think we've taken care of 2 and 3, so we just need to decide on `Neo4jOperations` and `Neo4jTemplate` now.

**Comment by mangrish** [comment]
Regarding point 1 we have agreed to keep the `Neo4jOperations` as deprecated and will be removed in a subsequent release. Does boot have any mechanism to load the bean on demand? So if someone has an `@Autowired Neo4jOperations` bean it could be built only then? This future proofs us to a degree with deprecation and our philosophy of encouraging users to use `Session` directly.

If not then feel free to add the bean and we can raise a ticket in a future release to remove this in future.

Thanks for your help @wilkinsona . 👍

**Comment by philwebb** [comment]
> Does boot have any mechanism to load the bean on demand? So if someone has an @Autowired Neo4jOperations bean it could be built only then?

No, we've not done anything like that in the past. I would suggest that we just keep the bean in 1.5 and then remove it in 2.0 when people have had a chance to see the deprecated warnings.

**Comment by mangrish** [comment]
Sounds good @philwebb. 👍  Let me know when this gets merged into your snapshot and I'll try it out.

**Comment by wilkinsona** [comment]
@mangrish The changes are now available in the latest 1.5.0 snapshots

**Comment by mangrish** [comment]
Thanks guys. I have an example here: https://github.com/neo4j-examples/sdn4-university/tree/spring-boot-1.5.

I have found one problem that I need your help with though. Will raise a new ticket for that.
